### PR TITLE
Open proper AO/AOCS segment files according to data volume

### PR DESCRIPTION
**Pull request by avamingli**
When inserting into AO/AOCS tables, we will open segment files equal to GUC gp_appendonly_insert_files anyway before. It will create uncessary empty segment files and increase total number of that, which leads to unexpected regression diffs related to total segfile number of AO/AOCS tables.

Open new segment file if there are enough data on the fly.

As we fix it, workarounds of gp_appendonly_insert_files in regression tests scripts could be resolved.
Adjust gp_appendonly_insert_files for some cases that have more data than gp_appendonly_insert_files_tuples_range to pass test like: inject fault and etc.

Authored-by: Zhang Mingli avamingli@gmail.com

closes: #215 

---

### Change logs

_Describe your change clearly, including what problem is being solved or what feature is being added._

_If it has some breaking backward or forward compatibility, please clary._

### Why are the changes needed?

_Describe why the changes are necessary._

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

_Please detail how the changes were tested, including manual tests and any relevant unit or integration tests._

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [ ] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [ ] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [ ] Document changes.
- [ ] Add tests for the change
- [ ] Pass `make installcheck`
- [ ] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to @cloudberrydb/dev team for review and approval when your PR is ready🥳

### CODE REVIEW

File: src/backend/access/aocs/aocsam_handler.c [hunk 1]
```diff
@@ -341,48 +352,41 @@ get_insert_descriptor(const Relation relation)
...
-		//* mark all insertDesc  placeholderInserted with false */
-        if (relationHasUniqueIndex(relation))
-        {
-            dlist_iter          iter;
-            dlist_foreach(iter, head)
-            {
-                AOCSInsertDesc insertDesc = (AOCSInsertDesc)dlist_container(AOCSInsertDescData, node, iter.cur);
-                insertDesc->placeholderInserted = false;
-            }
-        }
```
Comment (avamingli) [inline src/backend/access/aocs/aocsam_handler.c:377]: I remove this codes try to set `insertDesc->placeholderInserted = false`,  beccuase aocs_insert_init() palloc0, there is no need to fetch everyone and set it to false again.

@lss602726449  could you confirm that?

Comment (lss602726449) [reply src/backend/access/aocs/aocsam_handler.c:377]: Yes, it's my mistake. Thanks for correcting.

**Review by lss602726449 — APPROVED** [review]

**Review by my-ship-it — APPROVED** [review]
LGTM

**Review by Ray-Eldath — APPROVED** [review]

**Comment by my-ship-it** [comment]
https://github.com/cloudberrydb/cloudberrydb/issues/215

File: src/backend/access/aocs/aocsam_handler.c [hunk 2]
```diff
@@ -341,48 +352,41 @@ get_insert_descriptor(const Relation relation)
...
-	if (state->insertDesc->insertMultiFiles && state->insertDesc->range == gp_appendonly_insert_files_tuples_range)
+	if (state->insertMultiFiles && state->insertDesc->range == gp_appendonly_insert_files_tuples_range)
 	{
 		state->insertDesc->range = 0;
+
+		if (list_length(state->used_segment_files) < state->insertMultiFiles)
+		{
+			next = aocs_insert_init(relation, ChooseSegnoForWriteMultiFile(relation, state->used_segment_files));
+			dlist_push_tail(&state->head, &next->node);
+			state->used_segment_files = lappend_int(state->used_segment_files, next->cur_segno);
```
Comment (yjhjstz) [inline src/backend/access/aocs/aocsam_handler.c:387]: can `next->cur_segno` already in  `state->used_segment_files` ?

Comment (avamingli) [reply src/backend/access/aocs/aocsam_handler.c:387]: No, in principle. 
In what case it could be?

Comment (yjhjstz) [reply src/backend/access/aocs/aocsam_handler.c:387]: I don't know, but this case used_segment_files will increase , exhaust insertMultiFiles.

Comment (avamingli) [reply src/backend/access/aocs/aocsam_handler.c:387]: > used_segment_files will increase , exhaust insertMultiFiles.

Yeah, that's what we want to do, out goal is to ensure open a new variable segment file when needed, excluding  already opened ones, even by ourselves.

**Review by yjhjstz — APPROVED** [review]

**Pull request merged**